include/boost/windows_sspi/stream.hpp

```diff
@@ -33,14 +33,67 @@ namespace boost {
 namespace windows_sspi {
 
 // TODO: Move away from this file
-template <typename NextLayer, typename ConstBufferSequence> struct async_write_impl : boost::asio::coroutine {
+template <typename NextLayer>
+struct async_handshake_impl : boost::asio::coroutine {
+  async_handshake_impl(NextLayer& next_layer, detail::sspi_impl& sspi_impl)
+    : m_next_layer(next_layer)
+    , m_sspi_impl(sspi_impl) {
+  }
+
+  template <typename Self>
+  void operator()(Self& self, boost::system::error_code ec = {}, std::size_t length = 0) {
+    BOOST_ASIO_CORO_REENTER(*this) {
+      detail::sspi_handshake::state state;
+      while((state = m_sspi_impl.handshake()) != detail::sspi_handshake::state::done) {
+        if (state == detail::sspi_handshake::state::data_needed) {
+          // TODO: Use a fixed size buffer instead
+          m_input.resize(0x10000);
+          BOOST_ASIO_CORO_YIELD m_next_layer.async_read_some(net::buffer(m_input), std::move(self));
+          if (ec) {
+            self.complete(ec);
+            return;
+          }
+          m_sspi_impl.handshake.put({m_input.data(), m_input.data() + length});
+          m_input.clear();
+          continue;
+        }
+        if (state == detail::sspi_handshake::state::data_available) {
+          m_output = m_sspi_impl.handshake.get();
+          BOOST_ASIO_CORO_YIELD net::async_write(m_next_layer, net::buffer(m_output), net::transfer_exactly(m_output.size()), std::move(self));
+          if (ec) {
+            self.complete(ec);
+            return;
+          }
+          continue;
+        }
+        if (state == detail::sspi_handshake::state::error) {
```

Comment:
query: is it guaranteed that when you reach this line, at least one YIELD will have happened prior?
If not, you will cause the completion handler to execute on the wrong executor by calling self.complete(ec) directly here.
This can be mitigated by detecting that you are not in a continuation and:
```
if (!is_continuation_)
  BOOST_ASIO_CORO_YIELD 
    net::post(boost::beast::bind_front_handler(std::move(self), ec));
self.complete(ec);
```

Reply:
Even if it was guaranteed (and it isn't currently) I'd rather not rely on that being required to be guaranteed, as it could lead to (for me at least) very hard to debug errors.

I will look into it and can hopefully figure it out on my own, but is there some similar simple way to achieve this (I guess more or less "forcing" a yield) without using anything from `boost::beast`?

I don't think it makes too much sense to depend on `boost::beast` for this library, even if `boost::beast` is an amazing library of course :-)